flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/NonSpanningWrapper.java

```diff
@@ -290,4 +309,49 @@ public int read(byte[] b, int off, int len) {
 	public int read(byte[] b) {
 		return read(b, 0, b.length);
 	}
+
+	ByteBuffer wrapIntoByteBuffer() {
+		return segment.wrap(position, remaining());
+	}
+
+	int copyContentTo(byte[] dst) {
+		final int numBytesChunk = remaining();
+		segment.get(position, dst, 0, numBytesChunk);
+		return numBytesChunk;
+	}
+
+	/**
+	 * Copies the data and transfers the "ownership" (i.e. clears current wrapper).
+	 */
+	void transferTo(ByteBuffer dst) {
+		segment.get(position, dst, remaining());
+		clear();
+	}
+
+	Tuple2<DeserializationResult, Integer> getNextRecord(IOReadableWritable target) throws IOException {
+		int recordLen = readInt();
+		if (canReadRecord(recordLen)) {
+			return readInto(target);
+		} else {
+			return new Tuple2<>(PARTIAL_RECORD, recordLen);
+		}
+	}
```

Comment:
nit: by wrapper I meant a POJO not Tuple2, but this is probably fine.

However now I'm a bit worried about performance implications of this - it's a critical/hot path, and we are allocating new object on it.

Reply:
I'll create a POJO and will reuse the instance (it's a single thread).